### PR TITLE
update dependencies to latest version

### PR DESCRIPTION
**Pull request by humblec**
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

### CODE REVIEW

**Review by yonatankahana — COMMENTED** [review]
i don't see how it's related to #205. our current master report is:

```
2022-05-25T19:59:07.876+0300	INFO	Need to update DB
2022-05-25T19:59:07.876+0300	INFO	DB Repository: ghcr.io/aquasecurity/trivy-db
2022-05-25T19:59:07.876+0300	INFO	Downloading DB...
2022-05-25T19:59:16.305+0300	INFO	Detected OS: debian
2022-05-25T19:59:16.305+0300	INFO	Detecting Debian vulnerabilities...
2022-05-25T19:59:16.305+0300	INFO	Number of language-specific files: 1
2022-05-25T19:59:16.305+0300	INFO	Detecting gobinary vulnerabilities...

nfs-subdir-external-provisioner:master (debian 11.3)

Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)


nfs-subdir-external-provisioner (gobinary)

Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 1, CRITICAL: 0)

┌─────────────────────┬────────────────┬──────────┬────────────────────────────────────┬───────────────────────────────────┬───────────────────────────────────────────────────┐
│       Library       │ Vulnerability  │ Severity │         Installed Version          │           Fixed Version           │                       Title                       │
├─────────────────────┼────────────────┼──────────┼────────────────────────────────────┼───────────────────────────────────┼───────────────────────────────────────────────────┤
│ golang.org/x/crypto │ CVE-2022-27191 │ HIGH     │ v0.0.0-20210817164053-32db794688a5 │ 0.0.0-20220315160706-3147a52a75dd │ golang: crash in a golang.org/x/crypto/ssh server │
│                     │                │          │                                    │                                   │ https://avd.aquasec.com/nvd/cve-2022-27191        │
└─────────────────────┴────────────────┴──────────┴────────────────────────────────────┴───────────────────────────────────┴───────────────────────────────────────────────────┘
```

and this is the scan results for your branch:

```
2022-05-25T20:01:41.553+0300	INFO	Detected OS: debian
2022-05-25T20:01:41.553+0300	INFO	Detecting Debian vulnerabilities...
2022-05-25T20:01:41.553+0300	INFO	Number of language-specific files: 1
2022-05-25T20:01:41.553+0300	INFO	Detecting gobinary vulnerabilities...

nfs-subdir-external-provisioner:humblec (debian 11.3)

Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)


nfs-subdir-external-provisioner (gobinary)

Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 1, CRITICAL: 0)

┌─────────────────────┬────────────────┬──────────┬────────────────────────────────────┬───────────────────────────────────┬───────────────────────────────────────────────────┐
│       Library       │ Vulnerability  │ Severity │         Installed Version          │           Fixed Version           │                       Title                       │
├─────────────────────┼────────────────┼──────────┼────────────────────────────────────┼───────────────────────────────────┼───────────────────────────────────────────────────┤
│ golang.org/x/crypto │ CVE-2022-27191 │ HIGH     │ v0.0.0-20220214200702-86341886e292 │ 0.0.0-20220315160706-3147a52a75dd │ golang: crash in a golang.org/x/crypto/ssh server │
│                     │                │          │                                    │                                   │ https://avd.aquasec.com/nvd/cve-2022-27191        │
└─────────────────────┴────────────────┴──────────┴────────────────────────────────────┴───────────────────────────────────┴───────────────────────────────────────────────────┘
```

**Comment by humblec** [comment]
@yonatankahana not really into the referenced issue.. I removed it from the PR description.. ptal.. thanks.. 

The issue in the referred one is due to the fact that, we still hold `4.0.2` images in the registry.

**Review by yonatankahana — COMMENTED** [review]
have you tested it? it is compiling but the controller crash at startup.

```
➜  kubectl logs nfs-nfs-subdir-external-provisioner-84845d45cb-dvqpt
F0526 18:17:26.290834       1 controller.go:887] Error creating lock: endpoints lock is removed, migrate to endpointsleases
➜  kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.2", GitCommit:"8b5a19147530eaac9476b0ab82980b4088bbc1b2", GitTreeState:"clean", BuildDate:"2021-09-15T21:38:50Z", GoVersion:"go1.16.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.4", GitCommit:"e6c093d87ea4cbb530a7b2ae91e54c0842d8308a", GitTreeState:"clean", BuildDate:"2022-03-06T21:32:53Z", GoVersion:"go1.17.7", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by humblec** [comment]
Ok.. the leader election is on `endpoint` in the vendored controlller.. https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/controller/controller.go#L867 which has to be updated to `lease` or atleast `endpointslices`. let me get that in first.

**Comment by yonatankahana** [comment]
follow https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/pull/120

**Comment by yonatankahana** [comment]
@humblec 
since i've already tested the fix (https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/pull/120) with our repository, i can save you some time, even though they not released yet:

1. you will have change all the `/v6` to `/v8` in `go.mod` and all imports
2. `serverVersion.GitVersion` not required anymore in provisioner.go when calling `controller.NewProvisionController()`
3. the `ClusterRole` should have additional rules:

```yaml
  - apiGroups: ["coordination.k8s.io"]
    resources: ["leases"]
    verbs: ["get", "create", "update"]
```

hope i didn't miss anything :)

**Comment by humblec** [comment]
@yonatankahana thanks ;)... refreshed the PR with the changes mentioned in the above comment (https://github.com/kubernetes-sigs/nfs-subdir-external-provisioner/pull/206/commits/d581b9c2110494c77ab767ba362f84a2d9396751) + updating chart role binding to have the leases VERB.
However the linter fails [1] and complain it need version bump. so the same commit updated the chart version to `4.0.16` [1] but the linter is still not happy. I have added one more commit (https://github.com/kubernetes-sigs/nfs-subdir-external-provisioner/pull/206/commits/1e6a2c038d72437813a9f1208ae32e31accca274) on top of this to avoid version bump check in helm 

[1]
```
Linting chart 'nfs-subdir-external-provisioner => (version: "4.0.16", path: "charts/nfs-subdir-external-provisioner")'
Checking chart 'nfs-subdir-external-provisioner => (version: "4.0.16", path: "charts/nfs-subdir-external-provisioner")' for a version bump...
Error: Error linting charts: Error processing charts
Old chart version: 4.0.16
New chart version: 4.0.16
------------------------------------------------------------------------------------------------------------------------
 ✖︎ nfs-subdir-external-provisioner => (version: "4.0.16", path: "charts/nfs-subdir-external-provisioner") > Chart version not ok. Needs a version bump!
------------------------------------------------------------------------------------------------------------------------
```

File: charts/nfs-subdir-external-provisioner/Chart.yaml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 apiVersion: v1
-appVersion: 4.0.2
+appVersion: 4.0.16
```
Comment (yonatankahana) [inline charts/nfs-subdir-external-provisioner/Chart.yaml:2]: what should be bumped is `version`, not `appVersion`

**Comment by humblec** [comment]
@yonatankahana ptal.. thanks :+1:

**Comment by mengjiao-liu** [comment]
@humblec Are you still working on it？

**Comment by humblec** [comment]
> @humblec Are you still working on it？

@mengjiao-liu #258 should address it.. Thanks!

**Pull request closed without merging**